### PR TITLE
Fixes #24022 - Adds container health support to docker ps filter/format

### LINKED ISSUE #24022: [feature proposal] docker ps status/health UX change
```
$ docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS                              PORTS               NAMES
3c16d5f74e39        nginx:latest        "nginx -g 'daemon off"   3 seconds ago       Up 2 seconds                        80/tcp, 443/tcp     test.1.1d2acd2imm80y1zqevcnp8ory
b09df8d0e51e        hc_test:latest      "nginx -g 'daemon off"   4 minutes ago       Up Less than a second (unhealthy)   80/tcp, 443/tcp     hc.5.83rrgliqdfncxoutcp8w5jxgy
00b06c034c65        hc_test:latest      "nginx -g 'daemon off"   4 minutes ago       Up Less than a second (unhealthy)   80/tcp, 443/tcp     hc.3.8q3nhanfyavahabvu53cl3gdf
```

Since services and standalone containers live in the same environment it seems their output should be pretty uniform. For example, why the container at the top of the `docker ps` above has no health check under status isn't clear. I'm proposing that `STATUS` stays the same as it was pre-healthchecks. The health status should be under a new column,`HEALTH` (taken from inspect field name), which can be `starting`, `healthy`, `unhealthy` or `unknown`. `unknown` would be a new enum value for `Health` struct, which simply means there are no health checks so the health can't be determined. 

This _seems_ pretty straightforward looking through the code so if it sounds good I was going to attempt a PR.

### PR DESCRIPTION
**Pull request by jhorwit2**
Fixes #24022

**\- What I did**

Added support in `docker ps` to filter by health status.

**\- How I did it**

Added `health` filter to `docker ps` with values of `no healthcheck`, `starting`, `healthy` and `unhealthy`. 

**\- How to verify it**
1. Run some containers with and without health checks to test the feature.

```
$ docker run -d tutum/hello-world
```

```
$ docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS               NAMES
cd9abd6b523f        tutum/hello-world   "/bin/sh -c 'php-fpm "   2 seconds ago       Up 1 seconds        80/tcp              sad_panini

$ docker ps --filter "health=healthy"
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES

$ docker ps --filter "health=unhealthy"
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES

$ docker ps --filter "health=no healthcheck"
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS               NAMES
d1b7678e816f        tutum/hello-world   "/bin/sh -c 'php-fpm "   2 minutes ago       Up 2 minutes        80/tcp              pedantic_cray
```

Now lets add a container with a passing & failing check health check to make it fun

```
$ docker run --name failing --health-cmd "cat etc" --health-interval 1s -d tutum/hello-world
$ docker run --name passing --health-cmd "ls etc" --health-interval 1s -d tutum/hello-world

$ docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS                      PORTS               NAMES
dc232283a140        tutum/hello-world   "/bin/sh -c 'php-f..."   34 seconds ago      Up 33 seconds (unhealthy)   80/tcp              failing
d802ddbaa28c        tutum/hello-world   "/bin/sh -c 'php-f..."   42 seconds ago      Up 41 seconds (healthy)     80/tcp              passing
6ba7c51e7530        tutum/hello-world   "/bin/sh -c 'php-f..."   2 minutes ago       Up 2 minutes                80/tcp              tender_heisenberg

$ docker ps --filter "health=healthy"
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS                    PORTS               NAMES
d802ddbaa28c        tutum/hello-world   "/bin/sh -c 'php-f..."   56 seconds ago      Up 55 seconds (healthy)   80/tcp              passing

$ docker ps --health --filter "health=unhealthy"
CONTAINER ID        IMAGE               COMMAND                  CREATED              STATUS                          PORTS               NAMES
dc232283a140        tutum/hello-world   "/bin/sh -c 'php-f..."   About a minute ago   Up About a minute (unhealthy)   80/tcp              failing

$ docker ps --filter "health=no healthcheck"
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS               NAMES
6ba7c51e7530        tutum/hello-world   "/bin/sh -c 'php-f..."   3 minutes ago       Up 3 minutes        80/tcp              tender_heisenberg
```

**\- Description for the changelog**

Added container health support to docker ps filter

**\- A picture of a cute animal (not mandatory but encouraged)**

cute animal
Signed-off-by: Josh Horwitz horwitzja@gmail.com

### CODE REVIEW

**Comment by jhorwit2** [comment]
I haven't done docs yet because I wanted to make sure this approach was OK.

**Comment by jhorwit2** [comment]
Seems the win2lin build failed for a server config issue. ¯\_(ツ)_/¯

vendor failing is expected as per discussion with @thaJeztah bc of engine changes.

**Comment by thaJeztah** [comment]
Yes, the win2lin nodes currently have some issues; it's being looked into (so don't worry about that).

Thanks! Looking at this, I'm wondering if;

1. we need to add the "health" column by default (because we currently show it by default)
2. we need to add a special flag for this, or just leave it to users to use a custom template if they don't want the "health" column to show by default

Just some initial thoughts, so please don't change anything yet! Let's wait for other people to leave their thoughts :)

**Comment by jhorwit2** [comment]
@thaJeztah anyone in particular who would be good to take a gander at this? Was thinking it might be nice to include in 1.12 since that way it's introduced in 1 way instead of changing after GA.

**Comment by jhorwit2** [comment]
My commit is signed :(

**Comment by thaJeztah** [comment]
@jhorwit2 let me try and get some people to look at it, but not sure if it will make it for 1.12

ping @talex5 @stevvooe @tiborvass

**Comment by jhorwit2** [comment]
Understood, getting pretty close. Figured I'd try 👼

**Comment by talex5** [comment]
It does seem like "unhealthy" is important information that should appear by default, even if the other states aren't shown. Perhaps the client should add `(unhealthy)` to the status column if the health column isn't shown? Seems a big hacky, though.

**Comment by jhorwit2** [comment]
I agree health state is important, but only really for non-service containers. I say that because service tasks that are unhealthy wont stay in that state for long (10 seconds maybe) bc once it hits that state it's instantly sent a kill signal. I also didnt want to spam ps with "no healthcheck" for everyone with pre-1.12 containers. 

At least this way it is optional so users can add it to their conf format to see it, then later we can make it default based on feedback. If a lot of "hey i cant find the health of my container" tickets come in then that lets us know we should "promote" it to being default. 

Perhaps another option could be to show the health column when there is a container with a healthcheck.

**Comment by icecrime** [comment]
Definitely not a requirement for 1.12! Let me remove from milestone.

File: api/client/container/ps.go
Comment (thaJeztah) [inline api/client/container/ps.go]: This should be added as a filter instead of a separate flag

**Comment by thaJeztah** [comment]
This needs some more discussion; how are we going to handle showing/hiding columns; is that going to be based on the `--format` or, add a separate flag (e.g. `--columns` / `--show`);

current thought is to have both, but make them mutually exclusive (`--show` as an easy way, and `--format` as advanced way)

**Comment by jhorwit2** [comment]
I like the columns option. I always found `format` to feel a little advanced because you have to know about go templates / the underlying objects. 

What I could do is remove the `--health` flag so it will show up when you filter/format. That way we aren't introducing a new flag that would be useless in a future pr for `--columns` or `--show`

**Comment by justincormack** [comment]
@jhorwit2 that seems a good idea.

**Comment by thaJeztah** [comment]
I think it should not be automatically added when using `--filter` alone (filter and format are different things IMO), but allowing to include it in `--format` sounds good (open to suggestions though)

**Comment by cpuguy83** [comment]
@jhorwit2 Just discussed this with other maintainers.
Sounds like if we can just do the health status filter and nothing else in this PR we can get this moving.
If there's some other functionality it would be better to do it in a separate PR so it doesn't hold up the functionality we all seem to be +1 on.

LMK if you are still interested in carrying this.
Thanks! Looking forward to getting this in!

**Comment by jhorwit2** [comment]
@cpuguy83 Sounds good! I should be able to update this PR today.

**Comment by jhorwit2** [comment]
@cpuguy83 @thaJeztah that should do the trick 👍. Added an integration test + ps docs. I _think_ that covers everything.

**Comment by vdemeester** [comment]
Desgin LGTM :angel:

**Comment by cpuguy83** [comment]
Design LGTM as well.
FYI, engine-api has been merged back to docker/docker (under api/types), so no need to vendor now, just update directly.

Moving to code-review.

**Comment by cpuguy83** [comment]
Also needs a rebase

**Comment by jhorwit2** [comment]
@cpuguy83 I rebased. Still is a vendor change though.

**Comment by cpuguy83** [comment]
@jhorwit2 No need to use `engine-api`. Types live in `github.com/docker/docker/api/types`

**Comment by jhorwit2** [comment]
@cpuguy83 Oh, I misread your comment. (need more coffee). I'll update the PR.

**Comment by jhorwit2** [comment]
@cpuguy83 👍  should be good now.

File: container/health.go
Comment (cpuguy83) [inline container/health.go]: What's up with this change?

**Comment by cpuguy83** [comment]
One question, otherwise code LGTM.

File: container/health.go
Comment (jhorwit2) [inline container/health.go]: If you take a look at `daemon/health.go` specifically 

```
func (d *Daemon) initHealthMonitor(c *container.Container) {
	// If no healthcheck is setup then don't init the monitor
	if getProbe(c) == nil {
		return
	}

	// This is needed in case we're auto-restarting
	d.stopHealthchecks(c)

	if c.State.Health == nil {
		h := &container.Health{}
		h.Status = types.Starting
		c.State.Health = h
	}

	d.updateHealthMonitor(c)
}
```

You can see the status for a starting container gets set to the type.Starting enum.

Comment (jhorwit2) [inline container/health.go]: Also, it perhaps made more sense with the original PR when health was broken out into another column in `docker ps`. I'm open to going back, but I think it would be best to keep to 1 value.

Comment (cpuguy83) [inline container/health.go]: It makes sense, just want to make sure we aren't changing a user-facing value or if we are we know the ramifications.

Comment (jhorwit2) [inline container/health.go]: @cpuguy83 this is user facing in the sense it's displayed in `docker ps` currently. That's currently the only use of it.

Comment (talex5) [inline container/health.go]: The reason for the original code was that without it you just see the container status as e.g. "Up 20 seconds (starting)" which could be confusing because it's not obvious that it's the health check that's starting, or what the difference is between "Up" and "starting". With separate columns (as in the original patch), "starting" is clear on its own.

Comment (jhorwit2) [inline container/health.go]: @talex5 I agree. I have an idea I think everyone will like. updating.

**Comment by jhorwit2** [comment]
@talex5 @cpuguy83 I went back to original string `health: starting`. Thoughts on something like `Starting Health Check` over `Up <time> (health: starting)`

**Comment by jhowardmsft** [comment]
There's a problem with the test on RS1 - the test is hanging indefinitely.... Is there a race condition in the test? Haven't looked at the code though. Happened to notice as was watching the Jenkins server at the time and noticed a run taking an inordinate amount of time. Periodically I'm seeing a `docker inspect --format={{.State.Health.Status}} failing_container` process pop up in task manager every 5 to 10 seconds if that helps. Also in the daemon under test log, the last entry is `time="2016-09-08T18:29:04.226587400Z" level=warning msg="Unknown healthcheck type 'NONE' (expected 'CMD')"`, the timing of which lines up with the newly added test.

**Comment by jhorwit2** [comment]
@jhowardmsft I've never had any issues running the tests on my mac. So, it does appear the commonly used `waitForHealthStatus` function **could** run for forever if the container never changes Status. Wonder if there is an issue on windows showing up due to this test? I don't really have a way to test windows. The command being run before the wait call is just

```
out, _ = dockerCmd(c, "run", "--name=failing_container", "--health-cmd=cat etc", "--health-interval=1s", "-itd", "busybox")
```

File: integration-cli/docker_cli_ps_test.go
Comment (cpuguy83) [inline integration-cli/docker_cli_ps_test.go]: Can change this to `runSleepingContainer`

Comment (jhowardmsft) [inline integration-cli/docker_cli_ps_test.go]: ^^ :+1: The Windows busybox image can't be used interactively due to various reasons. So that container would just exit immediately. It sounds regardless, the wait shouldn't wait indefinitely and should have a reasonable timeout (1 min or so) for test failure.

**Comment by jhorwit2** [comment]
@jhowardmsft updated to use `runSleepingContainer`. Hopefully no more hanging tests 👼 .

**Comment by jhowardmsft** [comment]
LGTM from Windows perspective. Will leave to other maintainers to also approve/merge :smile:

**Comment by thaJeztah** [comment]
ping @cpuguy83 PTAL (I see some changes were made since your LGTM)

**Comment by cpuguy83** [comment]
TBH not really happy with where we are here where we have the daemon determining the presentation of the client.
@talex5 Why wouldn't health be an extra field on the ps API, and if need be we let the docker client combine health and run state?

**Comment by talex5** [comment]
@cpuguy83 I agree; was just copying the existing style, where the daemon combines the state (e.g. `Up`), the running time and the paused state into a single string before sending it back to the client.

**Comment by mlaventure** [comment]
@jhorwit2 this need a rebase.

ping @cpuguy83

**Comment by jhorwit2** [comment]
@mlaventure I've been waiting on a consensus on what should be done. What @cpuguy83 said is valid, but IMO out of scope for this diff. 

Is the approach in this diff fine or should other stuff be done?

**Comment by thaJeztah** [comment]
ping @cpuguy83 could you have a look when you're back from PTO? Let's see if we can get this into 1.13 (code freeze next week)

**Comment by mlaventure** [comment]
My 2cts:

I also agree that it may be a good idea to move presentation decision into the client, but that can be done in separate PR(s). Also it may not be that simple since we need to keep API compatibility and some fields are set as string (i.e. pre-processed by the daemon) on some of the endpoint.

As such, I'd vote to move along this PR as this filter would be quite useful.

**Comment by jhorwit2** [comment]
@mlaventure rebased

File: api/types/types.go
Comment (thaJeztah) [inline api/types/types.go]: just looking at this; having a space in this value could make it less easy to use (now I _must_ quote the filter), should we use (e.g.) "none", or "no-healthcheck" ?

**Comment by jhowardmsft** [comment]
The RS1 error looks legitimate:

```
14:48:00 ----------------------------------------------------------------------
14:48:00 FAIL: docker_cli_ps_test.go:230: DockerSuite.TestPsListContainersFilterHealth
14:48:00 
14:48:00 docker_cli_ps_test.go:265:
14:48:00     waitForHealthStatus(c, "passing_container", "starting", "healthy")
14:48:00 docker_cli_health_test.go:39:
14:48:00     c.Check(out, checker.Equals, prev)
14:48:00 ... obtained string = "unhealthy\n"
14:48:00 ... expected string = "starting\n"
14:48:00 
14:48:00 docker_cli_ps_test.go:269:
14:48:00     c.Assert(containerOut, checker.Equals, containerID, check.Commentf("Expected containerID %s, got %s for healthy filter, output: %q", containerID, containerOut, out))
14:48:00 ... obtained string = ""
14:48:00 ... expected string = "73f91a29e89367222962f1a0c725f93c84879997fd95f92cc98a5942a6831a69"
14:48:00 ... Expected containerID 73f91a29e89367222962f1a0c725f93c84879997fd95f92cc98a5942a6831a69, got  for healthy filter, output: ""
14:48:00 
14:48:04 
14:48:04 ----------------------------------------------------------------------
```

**Comment by jhorwit2** [comment]
:(
```
INFO: Testing against a local daemon
PASS: docker_cli_ps_test.go:230: DockerSuite.TestPsListContainersFilterHealth	8.489s
OK: 1 passed
PASS
---> Making bundle: .integration-daemon-stop (in bundles/1.13.0-dev/test-integration-cli)
++++ cat bundles/1.13.0-dev/test-integration-cli/docker.pid
+++ kill 8132
```

Locally it works. I'll try to take a look tomorrow.

**Comment by jhorwit2** [comment]
Oh, I think I know... It's the busybox `-it` windows issue.

**Comment by thaJeztah** [comment]
@jhorwit2 were you still looking into the failing windows test; need any help?

**Comment by jhorwit2** [comment]
@thaJeztah if anyone is able to help thatd be great :). Im traveling for work so im away from my only windows machine, which makes debugging hard.

File: api/types/types.go
Comment (tiborvass) [inline api/types/types.go]: 👍 for `none`

**Review by tiborvass — CHANGES_REQUESTED** [review]

**Comment by jhorwit2** [comment]
I should have time tomorrow on the flight back to take a look at the tests and update the const value 👍. Hopefully windows stops hating me 👼 .

**Comment by jhorwit2** [comment]
@thaJeztah 💥. all good. 

It seems `runSleepingContainer` + `-d` = breaks windows.

**Review by vdemeester — APPROVED** [review]
LGTM 🐸
/cc @tiborvass

**Review by tiborvass — APPROVED** [review]

**Comment by tiborvass** [comment]
LGTM

**Comment by vdemeester** [comment]
Moving it to docs-review
/cc @thaJeztah

**Review by thaJeztah — APPROVED** [review]
docs LGTM

**Comment by thaJeztah** [comment]
ping @albers @sdurrheimer for completion scripts :heart:

**Comment by vdemeester** [comment]
@jhorwit2 I think `man/docker-ps.1.md` also needs some updates 👼

**Comment by jhorwit2** [comment]
updated man page @vdemeester @thaJeztah

**Review by vdemeester — APPROVED** [review]
docs LGTM 🐯

**Pull request merged**
